P3732r2 (#43)
rarutyun/iso_cpp_papers: parallel_algorithms/P3732/numeric_ranges_algorithms.md

somehow).

Outcome: No objection to unanimous consent.

**Poll 3**: To associate an ad-hoc lambda with an identity, we prefer the option
`reduce(in, init, binary_operation([](auto lhs, auto rhs) { ... }, identity))` over the options
`reduce(in, init, [](auto lhs, auto rhs) { ... }, op_identity(identity))` or
`reduce(in, init, [](auto lhs, auto rhs) { ... }, identity)`.

SF | F | N | A | SA
---|---|---|---|---
4 | 3 | 0 | 0 | 0

Outcome: Strong consensus in favor.

**Poll 4**: We want the serial version of `reduce_into` (and variants) for consistency now.

Outcome: No objection to unanimous dissent.

**Poll 5**: We want (the parallel) `reduce_into` to take a range as output instead of an output
iterator (like all the other parallel range algorithms).

Outcome: No objection to unanimous consent.

**Poll 6**: We want projections for the algorithms proposed by P3732R1 "Numeric Range Algorithms".

SF | F | N | A | SA
---|---|---|---|---
0 | 0 | 2 | 4 | 2

Outcome: Strong consensus against.

# Motivation

C++20 added ranges version of algorithms. Unfortunately, only algorithms from `<algorithm>` and

> `<memory>` headers were added; not from `<numeric>` header. Also, all the added algorithms are
> non-parallel.

Reviewer: ```suggestion
`<memory>` headers were added, not from the `<numeric>` header. Also, parallel overloads of the added algorithms were omitted.
```

> Later C++23 added more algorithms to `ranges` namespace including `ranges::iota` to `<numeric>`
> header and `ranges::fold_` algorithm family, which is in `<algorithm>` header but could arguably be
> in `<numeric>` header as well, since it is very close to `reduce` or `accumulate`. No
> parallel algorithms in namespace `ranges` for C++23 timeframe, though.

Reviewer: ```suggestion
Later C++23 added more algorithms to `ranges` namespace.
This included adding `ranges::iota` to `<numeric>`
header and adding the `ranges::fold_` algorithm family.
While the latter are in the `<algorithm>` header, they could arguably have been added
to the `<numeric>` header as well, since they are very close to `reduce` or `accumulate`. C++23 did not add any
parallel algorithms in namespace `ranges`, though.
```

The good news is C++26 finally adds Parallel Range Algorithms (see [@P3179R9]) for the existing

> algorithms in `ranges` namespace, where applicable. Still we are missing `<numeric>` algorithms that
> are extremely useful for parallelism and important for HPC use cases, like `reduce` and `scan`.

Reviewer: They are important for all kinds of use cases where parallelism can help.  "HPC use cases" can sound a bit like "special stuff for those weird science people."

```suggestion
algorithms in `ranges` namespace, where applicable. Still, we are missing `<numeric>` algorithms that
are extremely useful for parallel computation, like `reduce` and `scan`.
```

Thus, this paper aims to address this gap by adding both parallel and non-parallel numeric range

> algorithms, which make the most sense from the authors perspective.

Reviewer: ```suggestion
algorithms.
```

Of course we add the algorithms that make sense for us to add.  We don't have to say that : - ) .

Finally, we propose parallel and non-parallel convenience wrappers:

* `ranges::sum` and `ranges::product` for special cases of `reduce` with addition and
  multiplication, respectively;
* `ranges::dot` for the special case of binary `transform_reduce` with transform `multiplies{}` and
  reduction `plus{}`; and
* `ranges::sum_into`, `ranges::product_into`, and `ranges::dot_into`
    (the "`_into`" versions of `sum`, `product`, and `dot`).

The following sections explain why we propose these algorithms and not others.
This relates to other aspects of the design besides algorithm selection,

> such as whether to include optional projection parameters. For more information please look at
> [](#scope-design-rationale).

Reviewer: ```suggestion
such as whether to include optional projection parameters. For more details please see
[this section](#scope-design-rationale).
```

                                             T init, F op);

template<forward_range R, forward_range OutR, class T = iter_value_t<I>,
         @_indirectly-binary-foldable_@<iterator<I>, T> F>
constexpr in_out_result<ranges::borrowed_iterator_t<R>, borrowed_iterator_t<O>>
  exclusive_scan(R&& result, OutR&& result_last,
                 T init, F op);

// Parallel overloads of exclusive scan

template<@_execution-policy_@ Ep, random_access_iterator I, sized_sentinel_for<I> S,
         random_access_iterator O, sized_sentinel_for<O> OutS,
         class T = iter_value_t<I>,
         @_indirectly-binary-foldable_@<I, T> F>
in_out_result<I, O> exclusive_scan(I first, S last, O result, OutS result_last,
                                   T init, F op);

template<@_execution-policy_@ Ep, @_sized-random-access-range_@ R,
         @_sized-random-access-range_@ OutR, class T = iter_value_t<I>,
         @_indirectly-binary-foldable_@<iterator<I>, T> F>
constexpr in_out_result<ranges::borrowed_iterator_t<R>, borrowed_iterator_t<O>>
  exclusive_scan(Ep&& exec, R&& result, OutR&& result_last,
                 T init, F op);
```

## Operation identity

> Based on our experience for C++17 Numeric Parallel Algorithms we allow to pass an optional
> operation identity (neutral to binary operation element) to Numeric Parallel Range Algorithms: to
> `*reduce` and `*scan` family, in particular.

Reviewer: ```suggestion
Based on our experience for C++17 Numeric Parallel Algorithms, we let users pass an optional
*identity* value for the binary operation (neutral to binary operation element) to the `*reduce` and `*scan` families of Numeric Parallel Range Algorithms.
```

Wouldn't this also include `reduce_into`?  If so, then `*reduce*` should be the family name.

> We propose that users are able to attach operation identity to a binary operation. The value can be
> known either at compile-time or at run-time.

Reviewer: ```suggestion
We propose that users are able to attach an identity value to a binary operation. The value can be
known either at compile time or at run time.
```
